Bio/SeqUtils/__init__.py

```diff
@@ -667,6 +667,46 @@ def calculate(self, sequence):
 
         return exp(cai_value / cai_length)
 
+    def optimize(self, sequence, seq_type="DNA", table=standard_dna_table):
+        """Return a new DNA sequence with preferred codons only."""
+        try:  # If seq record is provided, convert to sequence
+            sequence = sequence.seq
+        except AttributeError:  # not a  SeqRecord object
+            pass
+        seq = sequence.upper()
+        # Make dict with amino acids referencing preferred codons
+        pref_codons = {aminoacid: [] for aminoacid in table.protein_alphabet}
+        for codon, aminoacid in table.forward_table.items():
+            if self[codon] == 1.0:
+                if pref_codons[aminoacid] != []:
+                    message = f"{pref_codons[aminoacid]} and {codon} are equally preferred. Using {codon}"
+                    warnings.warn(message, RuntimeWarning)
+                pref_codons[aminoacid] = codon
+        for codon in table.stop_codons:
+            if self[codon] == 1.0:
+                pref_codons["*"] = codon
+        # Create amino acid sequence if DNA was provided
+        if seq_type == "DNA" or seq_type == "RNA":
+            aa_seq = translate(seq)
+        elif seq_type == "protein":
+            aa_seq = seq
+        else:
+            raise TypeError
+        # Un-translate in loop using only preferred codons
+        optimized = []
+        for aa in aa_seq:
+            try:
+                pref_codon = pref_codons[aa]
+            except KeyError:  # Keep stop codon from original sequence
+                if aa == "*":
+                    pref_codon = codon
```

Comment:
"Keep stop codon from original sequence"? This seems wrong. The `codon` variable you're referencing here is from the last iteration over `table.stop_codons` on line 685. Or, if there would be no stop codons in the table, then it would be from the last iteration over `table.forward_table.items()`. But, assuming there are stop codons in the table, then `"*"` can't throw a `KeyError` and  line 702 would be unreachable.

Reply:
Thanks! You're right, that was something I'd meant to remove.